src/Runtime.c

```diff
@@ -353,28 +355,76 @@ Term expand_ref(Loc def_idx) {
 
   Term root = term_offset_loc(nodes[0], offset);
 
-  // Unroll loop, better branch prediction
-  u32 i = 1;
-  for (; i + 7 < nodes_len; i += 8) {
-    set(i + offset, term_offset_loc(nodes[i], offset));
-    set(i + offset + 1, term_offset_loc(nodes[i + 1], offset));
-    set(i + offset + 2, term_offset_loc(nodes[i + 2], offset));
-    set(i + offset + 3, term_offset_loc(nodes[i + 3], offset));
-    set(i + offset + 4, term_offset_loc(nodes[i + 4], offset));
-    set(i + offset + 5, term_offset_loc(nodes[i + 5], offset));
-    set(i + offset + 6, term_offset_loc(nodes[i + 6], offset));
-    set(i + offset + 7, term_offset_loc(nodes[i + 7], offset));
-  }
-
-  // Remaining nodes
-  for (; i < nodes_len; i++) {
-    set(i + offset, term_offset_loc(nodes[i], offset));
+  if (nodes_len <= 32) {
+    for (u32 i = 1; i < nodes_len; i++) {
+      BUFF[offset + i] = term_offset_loc(nodes[i], offset);
+    }
+  } else {
+    #define STACK_BUFFER_SIZE 1024
+    Term stack_buffer[STACK_BUFFER_SIZE];
+    Term *offset_terms = (nodes_len - 1 <= STACK_BUFFER_SIZE) ? 
+                          stack_buffer : 
+                          malloc(sizeof(Term) * (nodes_len - 1));
+    
+    if (nodes_len - 1 > STACK_BUFFER_SIZE && !offset_terms) {
+      printf("expand_ref: failed to allocate memory\n");
+      exit(1);
+    }
+    
+    const u32 CHUNK_SIZE = 64; // Align to typical cache line size
+    for (u32 chunk_start = 1; chunk_start < nodes_len; chunk_start += CHUNK_SIZE) {
+      u32 chunk_end = (chunk_start + CHUNK_SIZE < nodes_len) ? 
+                      chunk_start + CHUNK_SIZE : nodes_len;
+      
+      if (chunk_end < nodes_len) {
+        __builtin_prefetch(&nodes[chunk_end], 0, 0);
+      }
+      
+      for (u32 i = chunk_start; i < chunk_end; i++) {
+        offset_terms[i-1] = term_offset_loc(nodes[i], offset);
+      }
+    }
+    
+    memcpy(&BUFF[offset + 1], offset_terms, sizeof(Term) * (nodes_len - 1));
```

Comment:
I'm pretty sure you're not allowed to access atomic values without using the atomic operations